### ISSUE TITLE
State of default constructed event should be re-considered

### ISSUE THREAD

**Issue by mkinsner** [KhronosGroup/SYCL-Docs#360]
This issue raises a reasonable (IMHO) use case of events:
https://github.com/Apress/data-parallel-CPP/issues/20

Default constructed events initialize to the ready state in SYCL 2020:

> Constructs an event that is immediately ready

This seems like a confusing and error prone design point.  Instead I'd expect the code in the linked issue to either work as the developer intended (wait and operate out of order) in the unsignaled interpretation, or to throw a synchronous exception that invalid events have been used on the initial `parallel_for` submission.

The current spec may make it easier to introduce program bugs than originally thought.  Should be discussed within the Khronos WG.

**Comment by gmlueck** [comment]
> Instead I'd expect the code in the linked issue to either work as the developer intended (wait and operate out of order) in the unsignaled interpretation

If the default created event was in the unsignaled state, then the user's code (linked above) would deadlock.  It would only make sense for the default-created event to be in the unsignaled state if we also provided some API which allowed the application to change it to "signaled".

> or to throw a synchronous exception that invalid events have been used on the initial parallel_for submission

This could make sense.  We did not think about having an "invalid" event when we wrote the SYCL 2020 spec.  Since then, we did create a DPC++ extension that adds the concept of an "invalid" event.  Most operations on an invalid event cause the operation to throw an exception.

FWIW, the use-case we had in mind for the current SYCL 2020 behavior is like:

```
event e; // Create event in "signaled" state
for (/*...*/) {
  e = q.single_task(e, [=] {/*...*/});
}
```

This makes it easy to create variable-length chains of dependent commands, rooted by a default-created event in the "signaled" state.  Obviously, the example above could be done more simply with an in-order queue, but you could imagine code like this being integrated with other commands that do not need to run sequentially.

**Comment by TApplencourt** [comment]
>If the default created event was in the unsignaled state, then the user's code (linked above) would deadlock.

Sorry, can you explain why?

**Comment by gmlueck** [comment]
> > If the default created event was in the unsignaled state, then the user's code (linked above) would deadlock.
> 
> Sorry, can you explain why?

I think this is an outcome of the common reference semantics.  Looking just at `e1`:

1. Line 11 default-constructs an event object, which is referred to by `e1`.  This event is in the unsignaled state.
2. Line 17 takes `e1` as a dependency parameter.  The `parallel_for` call makes a copy of `e1` internally, so it references the same object created in step 1.  This object is unsignaled, so the command does not execute.
3. The `paralel_for` on line 22 returns a _new_ event object.  Let's call it _Enew_.
4. The assignment operator on line 22 assigns the `e1` variable to this new object _Enew_.  The `e1` variable now refers to the _Enew_ object.  However, the copy taken in step 2 continues to refer to the default-constructed event object from step 1.

As a result, the `parallel_for` on line 17 never has its dependencies satisfied.  In fact, there is no way satisfy its dependencies because there is no API in SYCL that allows an application to change an existing event object to "signaled" state.

**Comment by TApplencourt** [comment]
This is what I was thinking; thanks all for confirming! I think the fact that the example will deadlock even if we start with unsignaled code is a good argument in favor of not supporting this use case.

**Comment by fraggamuffin** [comment]
Agree to close.

**Comment by fraggamuffin** [comment]
Wait for MK to review.

**Comment by gmlueck** [comment]
> but shouldn't we require a sync exception in these cases to reduce common programming errors? 

What specifically should cause a sync exception in your opinion?  Should we raise an exception whenever `parallel_for` is passed a dependent event that is already in "signaled" state?  That seems like a bad idea.

Are you proposing that `parallel_for` should raise an exception whenever it is passed a default-constructed event?  If we think this is a good idea, we should instead deprecate (or remove) the default-constructor for events because there would be no use for them anymore.

**Comment by mkinsner** [comment]
The second case is what I mean.  Removing default constructed events may be the right path, unless we ever think that we'll provide a way to mutate the state directly without the copy problem that you called out early in the issue.

**Comment by mkinsner** [comment]
Right.  I think that it's easy to write incorrect code through patterns like the OP example, which is orthogonal to the OP request.

The state of SYCL default constructed objects (`info::event_command_status::complete`) isn't what many developers will expect or remember.  It differs from for example OpenCL user events (that can be created apart from a submission API), which initialize to `CL_SUBMITTED`.

If we deem it too late to remove default constructed events, there are multiple other solutions to explore.

**Comment by gmlueck** [comment]
The default constructor for `event` has been around since at least SYCL 1.2.1, with its current semantic of being in the "complete" state.  I think we should not remove it without going through a deprecation period.

**Comment by Pennycook** [comment]
> Are you proposing that `parallel_for` should raise an exception whenever it is passed a default-constructed event? If we think this is a good idea, we should instead deprecate (or remove) the default-constructor for events because there would be no use for them anymore.

It just occurred to me while working on something else that having a default constructor for events is useful _because_ events have common reference semantics.  Isn't a default constructor necessary if we want to allow developers to write classes like the below?

```c++
struct S {
  void foo(sycl::queue q) {
    e = q.parallel_for(...);
  }
  sycl::event e;
};
```

I think that if we didn't have a default constructor, the class would have to do something gross like store a pointer to a `sycl::event`.  It couldn't store a `sycl::event&` because we don't have one when constructing `S`.  If I'm right, then removing the default constructor might have a significant impact on the structure of (existing) user code.
